src/resource.ts

```diff
@@ -0,0 +1,207 @@
+  public accessToken: string;
+  public expiresIn: number;
+  public expiresAt: number;
+  public tokenType: string;
+  public scope: [string];
+  public refreshToken: string;
+  public patient: string;
+
+  constructor(authToken: any) {
+    this.accessToken = authToken.access_token;
+    this.expiresIn = authToken.expires_in; // in seconds
+    this.expiresAt = authToken.expires_at
+      ? authToken.expires_at
+      : moment()
+          .add(this.expiresIn * 1000)
+          .valueOf();
+    this.patient = authToken.patient;
+    this.refreshToken = authToken.refresh_token;
+    this.scope = authToken.scope;
+    this.tokenType = authToken.token_type;
+  }
+
+  isExpired(): boolean {
+    return moment(this.expiresAt).isBefore(moment());
+  }
+
+  validate(): void {
+    if (
+      !(
+        this.accessToken &&
+        this.expiresIn &&
+        this.expiresAt &&
+        this.patient &&
+        this.refreshToken &&
+        this.scope &&
+        this.tokenType
+      )
+    ) {
+      throw new Error("Invalid authorization token.");
+    }
+  }
+}
+
+function sleep(time: number) {
+  return new Promise((resolve) => {
+    setTimeout(resolve, time);
+  });
+}
+
+function isRetryable(error: any) {
+  let retryOK = false;
+
+  if (
+    error.response &&
+    retrySettings.retryableCodes.includes(error.response.status)
+  ) {
+    if (error.request && error.request.path) {
+      retryOK = endsWith(error.request.path);
+    }
+  }
+  return retryOK;
+}
+
+function endsWith(path: string) {
+  let matched = false;
+  for (let i = 0; i < retrySettings.retryableFhirPath.length; i++) {
+    if (path.endsWith(retrySettings.retryableFhirPath[i])) {
+      matched = true;
+      break;
+    }
+  }
+  return matched;
+}
+
+async function doRetry(fhirUrl: string, config: any) {
+  const interval = retrySettings.initInterval;
+  const maxAttempts = retrySettings.maxAttempts;
+
+  let resp = null;
+
+  for (let i = 0; i < maxAttempts; i += 1) {
+    const waitInMilliSec = eval(retrySettings.backOffExpr);
+    await sleep(waitInMilliSec);
+    try {
+      resp = await axios.get(fhirUrl, config);
+      break;
+    } catch (error: any) {
+      if (error.response) {
+        resp = error.response;
+      }
+    }
+  }
+  return resp;
+}
+
+async function refreshAccessToken(
+  authToken: AuthorizationToken,
+  bb: BlueButton
+) {
+  const resp = await axios.post(
+    `${String(bb.baseUrl)}/v${bb.version}/o/token/`,
+    {},
+    {
+      auth: {
+        username: bb.clientId,
+        password: bb.clientSecret,
+      },
+      params: {
+        grant_type: "refresh_token",
+        client_id: bb.clientId,
+        refresh_token: authToken.refreshToken,
+      },
+    }
+  );
+
+  if (resp.status !== 200) {
+    const msg = `Failed to refresh access token, status: ${resp.status}, error: ${resp.data}.`;
+    throw new Error(msg);
+  }
+
+  return new AuthorizationToken(resp.data);
+}
+
+export async function fetchData(
```

Comment:
I don't think that this function should be so prescriptive on the url. It makes change more difficult in the future.

Reply:
@JFU-GIT I'm agreeing with Nick here. Thinking that you could change the `resourceType` to `resourcePath` where the string of the FHIR path is passed in. For example, "fhir/Coverage/" could be passed in. This way if a new resource is added in the future, it will still work with it.

Maybe some constants could be provided to make it nicer to the SDK user. 

For example:

const COVERAGE_RESOURCE_PATH = "fhir/Coverage/"